File: Project1/Go/BubbleSort.go
/*
 * Name: Abdelrahman Ahmed
 * Professor: Dr. Dai
 * Course: CS416 - Software Engineering
 * Date: Feb 17, 2026
 *
 * File: BubbleSort.go
 * Description:
 *   Implements Bubble Sort, a simple comparison-based algorithm that repeatedly swaps adjacent out-of-order elements.
 *   It runs in O(n^2) time in the worst case and is mainly useful for learning or very small inputs.
 */

package main

// BubbleSort sorts the slice in ascending order (in-place).
func BubbleSort(a []int) {
	n := len(a)
	for i := 0; i < n-1; i++ {
		swapped := false
		for j := 0; j < n-1-i; j++ {
			if a[j] > a[j+1] {
				a[j], a[j+1] = a[j+1], a[j]
				swapped = true
			}
		}
		if !swapped {
			return
		}
	}
}

File: Project1/Rust/BubbleSort.rs
/*

    Name: Diego Lara
    Class ID: 10
    Professor: Dr. Dai
    Course: CS416 - Software Engineering
    Date: Feb 20, 2026
    Bubble Sort Implementation in Rust

    Commands to run the code:
    rustc BubbleSort.rs
    ./BubbleSort
*/

fn bubble_sort<T: Ord>(arr: &mut [T]) {
	let n = arr.len();

	if n < 2 {
		return;
	}

	for i in 0..n {
		let mut swapped = false;
		for j in 0..(n - 1 - i) {
			if arr[j] > arr[j + 1] {
				arr.swap(j, j + 1);
				swapped = true;
			}
		}

		if !swapped {
			break;
		}
	}
}

fn main() {
	let mut numbers = vec![64, 34, 25, 12, 22, 11, 90];
	println!("Before sorting: {:?}", numbers);
	bubble_sort(&mut numbers);
	println!("After sorting:  {:?}", numbers);
}
